File: elixir/connect/connect.exs
defmodule Connect do
  @doc """
  Calculates the winner (if any) of a board
  using "O" as the white player
  and "X" as the black player
  """
  @spec result_for([String.t]) :: :none | :black | :white
  def result_for(board) do
    board =
      board
      |> Enum.map(&String.graphemes(&1))

    white_board = with_index(board)
    black_board =
      board
      |> Enum.zip
      |> Enum.map(&Tuple.to_list(&1))
      |> with_index

    white =
      white_board
      |> List.last
      |> Enum.filter(fn {player, _} -> player == "O" end)
      |> explore([], white_board)
      |> connected?

    black =
      black_board
      |> List.last
      |> Enum.filter(fn {player, _} -> player == "X" end)
      |> explore([], black_board)
      |> connected?

    cond do
      white -> :white
      black -> :black
      true -> :none
    end
  end

  def explore([], visited, _board), do: visited

  def explore([h|t], visited, board) do
    h
    |> find_next_steps(visited, board)
    |> case do
      [] -> explore(t, [h|visited], board)
      neighbors -> explore(neighbors ++ t, [h|visited], board)
    end
  end

  def connected?(visited) do
    visited
    |> Enum.any?(fn {_, {row, _}} -> row == 0 end)
  end

  def find_next_steps({color, {i, j}}, visited, board) do
    lookaround(board, i, j)
    |> Enum.filter(fn {player, _} -> player == color end)
    |> Enum.reject(fn pos -> pos in visited end)
  end

  def with_index(board) do
    rows =  length(board)
    cols =
      board
      |> List.first
      |> length

    for i <- 0..rows-1 do
      for j <- 0..cols-1 do
        {at(board, i, j), {i, j}}
      end
    end
  end

  @doc """
  Get the value at row i and column j
  """
  @spec at([list], integer, integer) :: any
  def at(board, i, j) do
    cond do
      i < 0 or j < 0 -> nil
      true ->
        board
        |> Enum.at(i)
        |> case do
          nil -> nil
          l -> Enum.at(l, j)
        end
    end
  end

  @doc """
  Lookaround six sides of the hexagon
  and get the value of the neighbors
  """
  @spec lookaround([list], integer, integer) :: [any]
  def lookaround(board, i, j) do
    [
      board |> at(i-1, j),    # top left
      board |> at(i-1, j+1),  # top right
      board |> at(i, j-1),    # left
      board |> at(i, j+1),    # right
      board |> at(i+1, j-1),  # bottom left
      board |> at(i+1, j),    # bottom right
    ]
    |> Enum.reject(&(&1 == nil))
  end
end

File: python/connect/connect.py
class ConnectGame:
    def __init__(self, board):
        self.board = [row.split() for row in board.split('\n')]

    @property
    def height(self):
        return len(self.board)

    @property
    def width(self):
        return len(self.board[0])

    def transpose(self):
        self.board = [list(row) for row in zip(*self.board)]

    def get_winner(self):
        if self.is_connected('O'):
            return 'O'
        self.transpose()
        if self.is_connected('X'):
            return 'X'
        return ''

    def is_connected(self, player):
        for pos in self._get_starting_points(player):
            if self._search(player, 0, pos, []):
                return True
        return False

    def _search(self, player, i=0, j=0, visited=[]):
        if i == self.height-1:
            return True

        visited.append((i, j))
        neighbors = self._get_neighbors(player, i, j, visited)

        for neighbor in neighbors:
            if self._search(player, *neighbor, visited):
                return True
        return False

    def _get_starting_points(self, player):
        return (j for j, p in enumerate(self.board[0])
                if p == player)

    def _get_neighbors(self, player, i, j, visited):
        neighbors = [
            (i-1, j),    # top left
            (i-1, j+1),  # top right
            (i, j-1),    # left
            (i, j+1),    # right
            (i+1, j-1),  # bottom left
            (i+1, j)     # bottom right
        ]
        return ((i, j) for i, j in neighbors
                if 0 <= i < self.height
                and 0 <= j < self.width
                and (i, j) not in visited
                and self.board[i][j] == player)
